File: scala/bob/src/main/scala/Bob.scala
object Bob {
  def response(statement: String): String =
    (isAsking(statement), isShouting(statement), isSilent(statement)) match {
      case (false, false, true) => "Fine. Be that way!"
      case (false, true, false) => "Whoa, chill out!"
      case (true, false, false) => "Sure."
      case (true, true, false) => "Calm down, I know what I'm doing!"
      case _ => "Whatever."
    }

  private def isAsking(statement: String): Boolean = statement.trim.endsWith("?")
  private def isShouting(statement: String): Boolean =
    statement.exists(_.isLetter) && statement == statement.toUpperCase
  private def isSilent(statement: String): Boolean = statement.trim.isEmpty
}

File: python/bob/bob.py
def is_silent(phrase):
    return not phrase.strip()


def is_shouting(phrase):
    return phrase.isupper()


def is_asking(phrase):
    return phrase.strip().endswith("?")


def response(phrase):
    """
    Bob responds to your phrase.
    """
    if is_silent(phrase):
        return "Fine. Be that way!"
    elif is_shouting(phrase) and is_asking(phrase):
        return "Calm down, I know what I'm doing!"
    elif is_shouting(phrase):
        return "Whoa, chill out!"
    elif is_asking(phrase):
        return "Sure."
    else:
        return "Whatever."
